File: next/services/pkg/jxold/npcres/KNpcTemplate.go
package npcres

import (
	"strconv"
	"strings"

	"github.com/JXGAME04/DUANGODOT/next/services/pkg/jxold/text"
)

// Template is one row of Settings/npcs.txt (KNpcTemplate of the old client).  The row index
// (header = 0) is the template id the map files (Npc_C.dat) and the zone use.
type Template struct {
	ID          int    `json:"id"`
	Name        string `json:"name"` // UTF-8 (the file stores TCVN3)
	Kind        int    `json:"kind"`
	Camp        int    `json:"camp"`
	Series      int    `json:"series"`
	ResType     string `json:"res"` // row of 人物类型.txt (KNpcResNode name)
	ArmorType   int    `json:"armor"`
	HelmType    int    `json:"helm"`
	WeaponType  int    `json:"weapon"`
	HorseType   int    `json:"horse"`
	RideHorse   bool   `json:"ride"`
	StandFrame  int    `json:"stand_frame"`
	StandFrame1 int    `json:"stand_frame1"`
	DeathFrame  int    `json:"death_frame"`
	WalkFrame   int    `json:"walk_frame"`
	RunFrame    int    `json:"run_frame"`
	HurtFrame   int    `json:"hurt_frame"`
	WalkSpeed   int    `json:"walk_speed"` // scene units per logic frame (KNpc::ServeMove)
	RunSpeed    int    `json:"run_speed"`
	Stature     int    `json:"stature"`
	// combat (KNpcTemplate::Init): the AttackSpeed column is the attack length in frames
	AttackFrame int `json:"attack_frame"`
	CastFrame   int `json:"cast_frame"` // the CastSpeed column: length of a non-melee skill
	HitRecover  int `json:"hit_recover"`
	ReviveFrame int `json:"revive_frame"`
	LifeParam   int `json:"life_param"` // raw *Param columns, placeholders when no level script runs
	MinDamage   int `json:"min_damage"`
	MaxDamage   int `json:"max_damage"`
	Defense     int `json:"defense"`
	// server side of the row (KNpcTemplate.cpp #ifdef _SERVER): what KNpcAI decides with
	AIMode       int              `json:"ai_mode"`       // 1..3 active, 4..6 passive, 0 none
	AIParam      [10]int          `json:"ai_param"`      // AIParam1..10 = m_AiParam[0..9]; AIParam10 defaults to 5
	AIMaxTime    int              `json:"ai_max_time"`   // frames between two decisions (default 25)
	VisionRadius int              `json:"vision_radius"` // default 40
	ActiveRadius int              `json:"active_radius"` // default 30
	Skills       [5]TemplateSkill `json:"skills"`        // slots 1..4 = Skill1..4 / Level1..4 (KSkillList::m_Skills); 0 unused
	// the LevelScript column and the raw cells KNpcTemplate::InitNpcLevelData hands to the script
	LevelScript string            `json:"level_script"`
	Cells       map[string]string `json:"cells"`
}

// LevelCells are the columns InitNpcLevelData reads as raw strings for the level script.
var LevelCells = []string{
	"ExpParam", "ExpParam1", "ExpParam2", "ExpParam3",
	"LifeParam", "LifeParam1", "LifeParam2", "LifeParam3",
	"LifeReplenish",
	"ARParam", "ARParam1", "ARParam2", "ARParam3",
	"DefenseParam", "DefenseParam1", "DefenseParam2", "DefenseParam3",
	"MinDamageParam", "MinDamageParam1", "MinDamageParam2", "MinDamageParam3",
	"MaxDamageParam", "MaxDamageParam1", "MaxDamageParam2", "MaxDamageParam3",
	"FireResist", "ColdResist", "LightResist", "PoisonResist", "PhysicsResist",
	"Level1", "Level2", "Level3", "Level4",
}

// TemplateSkill is one of the Skill1..4 / Level1..4 pairs.  A level cell "a|b" means
// a + b * npc level (GetNpcLevelData -> GetData of the level scripts); both cells must be
// present or KNpcTemplate::InitNpcLevelData leaves the slot empty.
type TemplateSkill struct {
	ID     int     `json:"id"`
	LevelA float64 `json:"level_a"`
	LevelB float64 `json:"level_b"`
}

// ParseTemplates reads npcs.txt.  Numbers are read like KTabFile::GetInteger (C atoi; an
// empty cell gives the old default, e.g. 15 frames).
func ParseTemplates(data []byte) []Template {
	tab := ParseTab(data)
	out := make([]Template, 0, tab.Height())
	num := func(row int, col string, def int) int {
		s := tab.GetByName(row, col)
		if strings.TrimSpace(s) == "" {
			return def
		}
		return Atoi(s)
	}
	for row := 1; row <= tab.Height(); row++ {
		t := Template{ID: row - 1}
		if row == 1 {
			out = append(out, t) // the header keeps id 0 unused, like the old direct indexing
			continue
		}
		t.Name = text.TCVN3ToUTF8([]byte(tab.Get(row, 1)))
		t.Kind = num(row, "Kind", 0)
		t.Camp = num(row, "Camp", 0)
		t.Series = num(row, "Series", 0)
		t.ResType = tab.GetByName(row, "NpcResType")
		t.ArmorType = num(row, "ArmorType", 0)
		t.HelmType = num(row, "HelmType", 0)
		t.WeaponType = num(row, "WeaponType", 0)
		t.HorseType = num(row, "HorseType", 0)
		t.RideHorse = num(row, "RideHorse", 0) != 0
		t.StandFrame = num(row, "StandFrame", 15)
		t.StandFrame1 = num(row, "StandFrame1", 15)
		t.DeathFrame = num(row, "DeathFrame", 15)
		t.WalkFrame = num(row, "WalkFrame", 15)
		t.RunFrame = num(row, "RunFrame", 15)
		t.HurtFrame = num(row, "HurtFrame", 15)
		t.WalkSpeed = num(row, "WalkSpeed", 5)
		t.RunSpeed = num(row, "RunSpeed", 10)
		t.Stature = num(row, "Stature", 0)
		t.AttackFrame = num(row, "AttackSpeed", 20)
		t.CastFrame = num(row, "CastSpeed", 20)
		t.HitRecover = num(row, "HitRecover", 0)
		t.ReviveFrame = num(row, "ReviveFrame", 2400)
		t.LifeParam = num(row, "LifeParam", 1)
		t.MinDamage = num(row, "MinDamageParam", 1)
		t.MaxDamage = num(row, "MaxDamageParam", 3)
		t.Defense = num(row, "DefenseParam", 0)
		t.AIMode = num(row, "AIMode", 0)
		for i := 0; i < 10; i++ {
			def := 0
			if i == 9 {
				def = 5
			}
			t.AIParam[i] = num(row, "AIParam"+strconv.Itoa(i+1), def)
		}
		t.AIMaxTime = num(row, "AIMaxTime", 25)
		t.VisionRadius = num(row, "VisionRadius", 40)
		t.ActiveRadius = num(row, "ActiveRadius", 30)
		for slot := 1; slot <= 4; slot++ {
			id := strings.TrimSpace(tab.GetByName(row, "Skill"+strconv.Itoa(slot)))
			level := strings.TrimSpace(tab.GetByName(row, "Level"+strconv.Itoa(slot)))
			if id == "" || level == "" {
				continue
			}
			a, b := parseLevel(level)
			t.Skills[slot] = TemplateSkill{ID: Atoi(id), LevelA: a, LevelB: b}
		}
		t.LevelScript = strings.ToLower(strings.TrimSpace(tab.GetByName(row, "LevelScript")))
		t.Cells = map[string]string{}
		for _, c := range LevelCells {
			if v := strings.TrimSpace(tab.GetByName(row, c)); v != "" {
				t.Cells[c] = v
			}
		}
		out = append(out, t)
	}
	return out
}

// parseLevel splits "a|b" (GetParam of the level scripts); a lone number is "a|0".
func parseLevel(s string) (float64, float64) {
	parts := strings.SplitN(s, "|", 2)
	a, _ := strconv.ParseFloat(strings.TrimSpace(parts[0]), 64)
	b := 0.0
	if len(parts) == 2 {
		b, _ = strconv.ParseFloat(strings.TrimSpace(parts[1]), 64)
	}
	return a, b
}

File: next/server/zone/src/KNpcTemplate.cpp
#include "jx/zone/KNpcTemplate.h"

#include <algorithm>
#include <cmath>
#include <cstdlib>
#include <fstream>

#include <nlohmann/json.hpp>

#include "jx/log.hpp"
#include "jx/zone/KScriptCache.h"

namespace jx::zone {

std::optional<KNpcTemplateSet> KNpcTemplateSet::load(const std::string& file, std::string* error)
{
    auto fail = [&](std::string why) -> std::optional<KNpcTemplateSet> {
        if (error) *error = std::move(why);
        return std::nullopt;
    };
    std::ifstream in(file);
    if (!in) return fail("cannot open " + file);
    nlohmann::json j;
    try {
        in >> j;
    } catch (const std::exception& e) {
        return fail(std::string("npcs.json: ") + e.what());
    }
    KNpcTemplateSet set;
    try {
        // a named copy: items() keeps a reference to it (a temporary would die under the loop)
        const nlohmann::json templates = j.value("templates", nlohmann::json::object());
        for (const auto& [key, t] : templates.items()) {
            KNpcTemplate tpl;
            tpl.id = static_cast<std::uint32_t>(std::stoul(key));
            tpl.name = t.value("name", "");
            tpl.kind = t.value("kind", 0);
            tpl.camp = t.value("camp", 4);
            tpl.series = t.value("series", 0);
            tpl.stature = t.value("stature", 0);
            tpl.stand_frame = t.value("stand_frame", 15u);
            tpl.stand_frame1 = t.value("stand_frame1", 15u);
            tpl.walk_frame = t.value("walk_frame", 15u);
            tpl.run_frame = t.value("run_frame", 15u);
            tpl.attack_frame = t.value("attack_frame", 20u);
            tpl.cast_frame = t.value("cast_frame", 20u);
            tpl.hurt_frame = t.value("hurt_frame", 10u);
            tpl.death_frame = t.value("death_frame", 12u);
            tpl.hit_recover = t.value("hit_recover", 0u);
            tpl.revive_frame = t.value("revive_frame", 2400u);
            tpl.life_param = t.value("life_param", 1u);
            tpl.min_damage = t.value("min_damage", 1u);
            tpl.max_damage = t.value("max_damage", 3u);
            tpl.defense = t.value("defense", 0u);
            tpl.walk_speed = t.value("walk_speed", 5);
            tpl.run_speed = t.value("run_speed", 10);
            tpl.ai_mode = t.value("ai_mode", 0);
            if (const auto ap = t.find("ai_param"); ap != t.end() && ap->is_array()) {
                int i = 0;
                for (const auto& v : *ap) {
                    if (i >= 10) break;
                    tpl.ai_param[i++] = v.is_number() ? v.get<int>() : 0;
                }
            }
            tpl.ai_max_time = t.value("ai_max_time", 25u);
            tpl.vision_radius = t.value("vision_radius", 40);
            tpl.active_radius = t.value("active_radius", 30);
            if (const auto sk = t.find("skills"); sk != t.end() && sk->is_array()) {
                int slot = 1;
                for (const auto& s : *sk) {
                    if (slot >= 5) break;
                    KNpcTemplateSkill& d = tpl.skills[slot++];
                    if (!s.is_object()) continue;
                    d.id = s.value("id", 0);
                    d.level_a = s.value("level_a", 0.0);
                    d.level_b = s.value("level_b", 0.0);
                    d.known = s.value("known", false);
                    d.attack_radius = s.value("attack_radius", 0);
                    d.melee = s.value("melee", false);
                    d.target_self = s.value("target_self", false);
                    d.style = s.value("style", 0);
                }
            }
            tpl.level_script = t.value("level_script", "");
            if (const auto c = t.find("cells"); c != t.end() && c->is_object()) {
                for (const auto& [col, v] : c->items()) {
                    if (v.is_string()) tpl.cells[col] = v.get<std::string>();
                }
            }
            set.templates_[tpl.id] = std::move(tpl);
        }
    } catch (const std::exception& e) {
        return fail(std::string("npcs.json fields: ") + e.what());
    }
    log::info("npc", "npc templates loaded", {log::kv("file", file), log::kv("count", set.templates_.size())});
    return set;
}

const KNpcTemplate* KNpcTemplateSet::find(std::uint32_t id) const
{
    const auto it = templates_.find(id);
    return it == templates_.end() ? nullptr : &it->second;
}

namespace {

// KTabFile::GetFloat: atof of the cell, the default for an empty cell.
double cell_float(const KNpcTemplate& t, const std::string& column, double def)
{
    const std::string s = t.cell(column);
    if (s.empty()) return def;
    return std::strtod(s.c_str(), nullptr);
}

std::uint32_t to_u32(double v) noexcept
{
    return v <= 0 ? 0u : static_cast<std::uint32_t>(v);   // the old int members truncate
}

// GetNpcLevelDataFromScript(pScript, nSeries, name, nLevel, p1, p2, p3): GetNpcKeyData
int key_data(KLuaScript& s, int series, int level, const char* name, double p1, double p2, double p3)
{
    const auto v = s.call_number("GetNpcKeyData", {static_cast<double>(series), static_cast<double>(level), std::string(name), p1, p2, p3});
    return v ? static_cast<int>(*v) : 0;
}

// GetNpcLevelDataFromScript(pScript, nSeries, name, nLevel, szParam): GetNpcLevelData; an empty
// cell is 0 without a call.
int level_data_str(KLuaScript& s, int series, int level, const char* name, const std::string& cell)
{
    if (cell.empty()) return 0;
    const auto v = s.call_number("GetNpcLevelData", {static_cast<double>(series), static_cast<double>(level), std::string(name), cell});
    return v ? static_cast<int>(*v) : 0;
}

// nParam * GetNpcKeyData(...) / 100.0 for one stat group (ExpParam, ExpParam1..3 ...)
std::uint32_t percent_of(KLuaScript& s, const KNpcTemplate& t, int series, int level, const char* group, const char* name)
{
    const std::string g(group);
    const double param = cell_float(t, g, 1);
    const double p1 = cell_float(t, g + "1", 0);
    const double p2 = cell_float(t, g + "2", 0);
    const double p3 = cell_float(t, g + "3", 0);
    return to_u32(param * key_data(s, series, level, name, p1, p2, p3) / 100.0);
}

} // namespace

KNpcLevelData KNpcTemplateSet::level_data(const KNpcTemplate& t, int level, int series, KScriptCache* scripts)
{
    KNpcLevelData d;
    level = std::max(1, level);
    // placeholders: what the zone simulated with before the level scripts
    d.life_max = std::max(10u, std::max(1u, t.life_param) * static_cast<std::uint32_t>(level));
    d.min_damage = std::max(1u, t.min_damage);
    d.max_damage = std::max(d.min_damage, t.max_damage);
    d.attack_rating = 100;
    d.defend = t.defense;
    for (int slot = 1; slot < 5; ++slot) {
        if (t.skills[slot].id > 0) {
            d.skill_level[slot] = static_cast<int>(std::floor(t.skills[slot].level_a + t.skills[slot].level_b * level));
        }
    }
    if (scripts == nullptr) return d;
    KLuaScript* s = t.level_script.empty() ? nullptr : scripts->get(t.level_script);
    if (s == nullptr) s = scripts->get(KScriptCache::kNpcLevelScript);   // g_pNpcLevelScript
    if (s == nullptr) return d;

    // KNpcTemplate::InitNpcLevelData, in its order
    for (int slot = 1; slot < 5; ++slot) {
        const std::string name = "Level" + std::to_string(slot);
        const std::string cell = t.cell(name);
        if (t.skills[slot].id > 0 && !cell.empty()) {
            d.skill_level[slot] = level_data_str(*s, series, level, name.c_str(), cell);
        } else {
            d.skill_level[slot] = 0;
        }
    }
    d.exp = percent_of(*s, t, series, level, "ExpParam", "Exp");
    d.life_max = percent_of(*s, t, series, level, "LifeParam", "Life");
    if (d.life_max == 0) d.life_max = 100;
    d.life_replenish = level_data_str(*s, series, level, "LifeReplenish", t.cell("LifeReplenish"));
    d.attack_rating = percent_of(*s, t, series, level, "ARParam", "AR");   // "AR", not "AttackRating": most scripts fall through to the quadratic
    if (d.attack_rating == 0) d.attack_rating = 100;
    d.defend = percent_of(*s, t, series, level, "DefenseParam", "Defense");
    d.min_damage = percent_of(*s, t, series, level, "MinDamageParam", "MinDamage");
    d.max_damage = percent_of(*s, t, series, level, "MaxDamageParam", "MaxDamage");
    d.fire_resist = level_data_str(*s, series, level, "FireResist", t.cell("FireResist"));
    d.cold_resist = level_data_str(*s, series, level, "ColdResist", t.cell("ColdResist"));
    d.light_resist = level_data_str(*s, series, level, "LightResist", t.cell("LightResist"));
    d.poison_resist = level_data_str(*s, series, level, "PoisonResist", t.cell("PoisonResist"));
    d.physics_resist = level_data_str(*s, series, level, "PhysicsResist", t.cell("PhysicsResist"));
    d.from_script = true;
    return d;
}

} // namespace jx::zone
